turbine/src/quic_endpoint.rs

```diff
@@ -0,0 +1,429 @@
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+) -> Result<(Endpoint, AsyncSender<(SocketAddr, Bytes)>, JoinHandle<()>), Error> {
+    let (cert, key) = new_self_signed_tls_certificate(keypair, address)?;
+    let server_config = new_server_config(cert.clone(), key.clone())?;
+    let client_config = new_client_config(cert, key)?;
+    let mut endpoint = {
+        let _guard = runtime.enter();
+        Endpoint::new(
+            EndpointConfig::default(),
+            Some(server_config),
+            socket,
+            TokioRuntime,
+        )?
+    };
+    endpoint.set_default_client_config(client_config);
+    let cache = Arc::<RwLock<ConnectionCache>>::default();
+    let (client_sender, client_receiver) = tokio::sync::mpsc::channel(CLIENT_CHANNEL_CAPACITY);
+    let server_task = runtime.spawn(run_server(endpoint.clone(), sender.clone(), cache.clone()));
+    let client_task = runtime.spawn(run_client(endpoint.clone(), client_receiver, sender, cache));
+    let run_endpoint = move || {
+        let task = futures::future::try_join(server_task, client_task);
+        if let Err(err) = runtime.block_on(task) {
+            error!("runtime.block_on(task): {err:?}");
+        }
+    };
+    let join_handle = std::thread::Builder::new()
+        .name(String::from("solTurbineQuic"))
+        .spawn(run_endpoint)?;
+    Ok((endpoint, client_sender, join_handle))
+}
+
+fn new_server_config(cert: Certificate, key: PrivateKey) -> Result<ServerConfig, rustls::Error> {
+    let mut config = rustls::ServerConfig::builder()
+        .with_safe_defaults()
+        .with_client_cert_verifier(Arc::new(SkipClientVerification {}))
+        .with_single_cert(vec![cert], key)?;
+    config.alpn_protocols = vec![ALPN_TURBINE_PROTOCOL_ID.to_vec()];
+    let mut config = ServerConfig::with_crypto(Arc::new(config));
+    config
+        .transport_config(Arc::new(new_transport_config()))
+        .use_retry(true)
+        .migration(false);
+    Ok(config)
+}
+
+fn new_client_config(cert: Certificate, key: PrivateKey) -> Result<ClientConfig, rustls::Error> {
+    let mut config = rustls::ClientConfig::builder()
+        .with_safe_defaults()
+        .with_custom_certificate_verifier(Arc::new(SkipServerVerification {}))
+        .with_single_cert(vec![cert], key)?;
+    config.enable_early_data = true;
+    config.alpn_protocols = vec![ALPN_TURBINE_PROTOCOL_ID.to_vec()];
+    let mut config = ClientConfig::new(Arc::new(config));
+    config.transport_config(Arc::new(new_transport_config()));
+    Ok(config)
+}
+
+fn new_transport_config() -> TransportConfig {
+    let mut config = TransportConfig::default();
+    config
+        .max_concurrent_bidi_streams(VarInt::from(0u8))
+        .max_concurrent_uni_streams(VarInt::from(0u8))
+        .initial_max_udp_payload_size(INITIAL_MAX_UDP_PAYLOAD_SIZE);
+    config
+}
+
+async fn run_server(
+    endpoint: Endpoint,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
```

Comment:
How about adding an exit flag for aborting the server? Or do you intend to close everything directly through the `Endpoint`?

Reply:
Gotcha, that works